src/types/packet.ts

```diff
@@ -4,53 +4,108 @@ import {
 import { Position } from "./common";
 import { ViewGraph } from "./graphs/viewgraph";
-import { Layer } from "../types/devices/layer";
+import { Layer } from "./layer";
 //import { EchoMessage } from "../packets/icmp";
-import { DeviceId, isRouter, isSwitch } from "./graphs/datagraph";
-import { EthernetFrame } from "../packets/ethernet";
-import { IPv4Packet } from "../packets/ip";
+import { DataGraph, DeviceId } from "./graphs/datagraph";
+import { EthernetFrame, IP_PROTOCOL_TYPE } from "../packets/ethernet";
+import { ICMP_PROTOCOL_NUMBER, IPv4Packet } from "../packets/ip";
+import { GlobalContext } from "../context";
+import { DataRouter, DataSwitch } from "./data-devices";
+import {
+  ICMP_REPLY_TYPE_NUMBER,
+  ICMP_REQUEST_TYPE_NUMBER,
+  IcmpPacket,
+} from "../packets/icmp";
 
 const contextPerPacketType: Record<string, GraphicsContext> = {
   IP: circleGraphicsContext(Colors.Green, 0, 0, 5),
   "ICMP-8": circleGraphicsContext(Colors.Red, 0, 0, 5),
   "ICMP-0": circleGraphicsContext(Colors.Yellow, 0, 0, 5),
+  EMPTY: circleGraphicsContext(Colors.Grey, 0, 0, 5),
   TCP: circleGraphicsContext(Colors.Hazel, 0, 0, 5), // for HTTP
 };
 
 const highlightedPacketContext = circleGraphicsContext(Colors.Violet, 0, 0, 6);
 
-export class Packet extends Graphics {
-  viewgraph: ViewGraph;
+export interface PacketLocation {
+  startId: DeviceId;
+  endId: DeviceId;
+  currProgress: number;
+}
 
-  private speed = 100;
-  private progress = 0;
-  private currentEdge: Edge;
-  private currentStart: number;
-  private type: string;
-  private rawPacket: EthernetFrame;
+interface PacketContext {
+  type: string;
+  layer: Layer;
+}
 
-  constructor(viewgraph: ViewGraph, type: string, rawPacket: EthernetFrame) {
-    super();
+function packetIsVisible(
+  _: ViewGraph | DataGraph,
+  isVisible: boolean,
+): _ is ViewGraph {
+  return isVisible;
+}
 
-    this.viewgraph = viewgraph;
-    this.type = type;
+function packetContext(frame: EthernetFrame): PacketContext {
+  if (frame.payload.type() === IP_PROTOCOL_TYPE) {
+    const datagram = frame.payload as IPv4Packet;
+    if (datagram.payload.protocol() === ICMP_PROTOCOL_NUMBER) {
+      const packet = datagram.payload as IcmpPacket;
+      if (packet.type === ICMP_REQUEST_TYPE_NUMBER) {
+        return { type: "ICMP-8", layer: Layer.Network };
+      }
+      if (packet.type === ICMP_REPLY_TYPE_NUMBER) {
+        return { type: "ICMP-0", layer: Layer.Network };
+      }
+    }
+  }
+  return { type: "EMPTY", layer: Layer.Link };
+}
 
+export class Packet extends Graphics {
+  packetId: string;
+  protected speed = 100;
+  protected progress = 0;
+  protected currStart: DeviceId;
+  protected currEnd: DeviceId;
+  protected graph: ViewGraph | DataGraph;
+  protected type: string;
+  protected rawPacket: EthernetFrame;
+  ctx: GlobalContext;
+  belongingLayer: Layer;
+
+  constructor(
+    ctx: GlobalContext,
+    graph: ViewGraph | DataGraph,
+    rawPacket: EthernetFrame,
+    isVisible: boolean,
```

Comment:
We could replace the `isVisible` parameter with the result of `graph instanceof ViewGraph`.